README.md
# DateInRegion

[![Build Status](https://travis-ci.org/Hout/DateInRegion.svg?branch=master)](https://travis-ci.org/Hout/DateInRegion)
[![Version](https://img.shields.io/cocoapods/v/DateInRegion.svg?style=flat)](http://cocoapods.org/pods/DateInRegion)
[![License](https://img.shields.io/cocoapods/l/DateInRegion.svg?style=flat)](http://cocoapods.org/pods/DateInRegion)
[![Platform](https://img.shields.io/cocoapods/p/DateInRegion.svg?style=flat)](http://cocoapods.org/pods/DateInRegion)

DateInRegion is a wrapper around NSDate that exposes the properties of NSDateComponents, NSCalendar, NSTimeZone, NSLocale and NSDateFormatter. The best way to see it is as a localised (locale) moment (date) in a time zone (timeZone) within a calendrical system (calendar). The intention is to replace your occurrence of NSDate with DateInRegion and get the same functionality plus lots of local date/calendar/time zone/formatter goodies. Thus offering date functions with a flexibility that I was looking for when creating this library.

An extension is available for NSDate with which enables you to map an NSDate object to a DateInRegion object by adding the region.

### Features

- Use the object as an NSDate. I.e. as an absolute time. 
- Offers many NSDate, NSCalendar, NSDateFormatter & NSDateComponent vars & methods
- Initialise a date with any combination of components
- DateInRegion is immutable, so thread safe. It contains a constructor to easily create new `DateInRegion` occurrences with some properties adjusted.
- Implements date addition and subtraction operators with date components. E.g. `date + 2.days`
- Default date is `NSDate()`
- Default calendar is `NSCalendar.currentCalendar()`
- Default time zone is `NSTimeZone.defaultTimeZone()`
- Default locale is `NSLocale.currentLocale()`
- Contains a date (NSDate), a calendar (NSCalendar), a locale (NSLocale) and a timeZone (NSTimeZone) property
- Implements the ``Equatable`` & ``Comparable`` protocols betwen dates with operators. E.g. `==, !=, <, >, <=, >=`
- Implements the ``Hashable`` protocol so the date can be used as a key in a Dictionary.

### Examples
Check out the playground:

```swift
import DateInRegion

// First create some regions
let local = DateRegion()
let netherlands = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "Europe?Amsterdam", localeID: "nl_NL")
let india = DateRegion(calendarID: NSCalendarIdentifierIndian, timeZoneID: "IST", localeID: "en_IN")
let china = DateRegion(calendarID: NSCalendarIdentifierChinese, timeZoneID: "Asia/Shanghai", localeID: "zn_CH")

// Initialisers
// Create a new date object with the current date & time alike NSDate()
let now = DateInRegion()

// Create a determined date
let newDate = DateInRegion(year: 2011, month: 2, day: 11)!

// Create a determined date in a different time zone
let newYork = DateRegion(timeZoneID: "EST")
DateInRegion(fromDate: newDate, hour: 14, region: newYork)!

// Mind that default values for DateInRegion(year etc) are taken from the reference date,
// which is 1 January 2001, 00:00:00.000 in your default time zone and against your current calendar.

// Week oriented initiailisations are also possible: first week of 2016:
let weekDate = DateInRegion(yearForWeekOfYear: 2016, weekOfYear: 1, weekday: 1)!
// In Europe this week starts in 2015 despite the year for the week that is 2016.
// That is because the Thursday of this week is in 2016 as specified by ISO 8601

// Or get date components
now.hour
now.day
now.month
now.year

// NSDate conversions
let nsDate = NSDate()
nsDate.inRegion(newYork).toString()
nsDate.inRegion(india).toString()
nsDate.inRegion(china).toString()


// Equations
// DateInRegion conforms to the Equatable protocol. I.e. you can compare with == for equality.
let newDate2 = DateInRegion(fromDate: newDate)
let newDate3 = DateInRegion(fromDate: newDate, hour: 9)!
newDate == newDate
newDate == newDate2
newDate == newDate3

// For equal date values, you should use x.isEqualToDate(y) note that this compares the moment not the region

// Comparisons
// DateInRegion conforms to the Comparable protocol. I.e. you can compare with <. <=, ==, >=, >
newDate > newDate2
newDate >= newDate2
newDate == newDate2
newDate != newDate2
newDate <= newDate2
newDate < newDate2

// Mind that comparisons takes the absolute time from the date property into account.
// regions (calendars, time zones, locales, have no effect on the comparison results.
let date1 = DateInRegion(year: 2000, month: 1, day: 1, hour: 14, region: netherlands)!
let date2 = (date1 + 1.hours)!
let date3 = (date2 - 2.weeks)!

// date1 = 14:00 UTC
// date2 = 15:00 UTC
date1 > date2
date1 < date2

let indiaDate1 = DateInRegion(fromDate: date1, region: india)

// indiaDate1 = 18:30 IST
// date1 = 14:00 UTC
// date2 = 9:00 CST
indiaDate1 > date2
indiaDate1 < date2

indiaDate1.isEqualToDate(date1)
// QED: same outcome!

// Collections
// Use as a key in a dictionary
var birthdays = [DateInRegion: String]()
birthdays[DateInRegion(year: 1997, month: 5, day: 7)!] = "Jerry"
birthdays[DateInRegion(year: 1999, month: 12, day: 14)!] = "Ken"
birthdays[DateInRegion(year: 1990, month: 12, day: 5)!] = "Sinterklaas"
birthdays.sort({ (a: (date: DateInRegion, String), b: (date: DateInRegion, String)) -> Bool in
return a.date < b.date
})

// Display & string conversion
// DateInRegion conforms to the ConvertString protocol
indiaDate1.description

// DateInRegion conforms to the DebugConvertString protocol
indiaDate1.debugDescription

// Create some more regions to illustrate conversions & string handling
let tibetan = DateRegion(calendarID: NSCalendarIdentifierIndian, timeZoneID: "IST", localeID: "bo_IN")
let dubai = DateRegion(calendarID: NSCalendarIdentifierIslamic, timeZoneID: "GST", localeID: "ar_AE")
let newZealand = DateRegion(calendarID: NSCalendarIdentifierGregorian, localeID: "en_NZ", timeZoneID: "Pacific/Auckland")
let israel = DateRegion(calendarID: NSCalendarIdentifierHebrew, timeZoneID: "Asia/Jerusalem", localeID: "he_IL")
let magadan = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "Asia/Magadan", localeID: "ru_RU")
let thailand = DateRegion(calendarID: NSCalendarIdentifierBuddhist, timeZoneID: "Asia/Bangkok", localeID: "th_TH")
let japan = DateRegion(calendarID: NSCalendarIdentifierBuddhist, timeZoneID: "Asia/Tokyo", localeID: "ja_JP")
let unalaska = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "AKST", localeID: "en_US")
let utc = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "UTC", localeID: "en_US_POSIX")


// Conversions
let unalaskaDate = now.inRegion(unalaska)
let netherlandsDate = now.inRegion(netherlands)
let newYorkDate = now.inRegion(newYork)
let indiaDate = now.inRegion(india)
let tibetanDate = now.inRegion(tibetan)
let dubaiDate = now.inRegion(dubai)
let israelDate = now.inRegion(israel)
let chinaDate = now.inRegion(china)
let newZealandDate = now.inRegion(newZealand)
let magadanDate = now.inRegion(magadan)
let japanDate = now.inRegion(japan)
let thailandDate = now.inRegion(thailand)
let utcDate = now.inRegion(utc)

// Or convert to local
let japanDate2 = DateInRegion(year: 2010, month: 4, day: 5, hour: 16, region: japan)!
let localDate = japanDate2.inLocalRegion()

// Now look in regional format
unalaskaDate.toString()
netherlandsDate.toString()
newYorkDate.toString()
indiaDate.toString()
tibetanDate.toString()
dubaiDate.toString()
israelDate.toString()
chinaDate.toString()
magadanDate.toString()
japanDate.toString()
thailandDate.toString()
newZealandDate.toString()
utcDate.toString()
japanDate2.toString()
localDate.toString()

// Other string formats: with styles
newYorkDate.toString(dateStyle: .LongStyle)
newYorkDate.toString(timeStyle: .LongStyle)
newYorkDate.toString(dateStyle: .ShortStyle, timeStyle: .MediumStyle)

// Other string formats: with Unicode abbreviations 
// http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Format_Patterns
newYorkDate.toString("yy-MM-dd")
newYorkDate.toString("yyyy-DDD")
newYorkDate.toString("yy.ww.e")
newYorkDate.toString("HH:mm:ss.SSS")
newYorkDate.toString("hh:mm:ss a z")


```

## Usage

To run the example project, clone the repo, and run `pod install` from the Example directory first.

## Requirements

iOS 9 is required.
The test code requires Quick and Nimble to be implemented. You can do so by running ``pod install`` in the ``Examples`` folder.

## Installation

### Cocoapods
DateInRegion is available through [CocoaPods](http://cocoapods.org). To install
it, simply add the following line to your Podfile:

```ruby
pod "DateInRegion"
```
### Manual
Get a git clone and add ``DateInRegion.swift`` to your project.

```shell
git clone https://github.com/Hout/DateInRegion.git
```

## Design Decisions
I have taken the following decisions when setting up the library. I welcome  feedback on them.

Decision | Rationale
------------- | -------------
Primary function of `DateInRegion` is to be able to switch and convert dates between regions. Not to deal with dates in the local region, although it can be used for that.|That is just the way I needed it :-)
`DateInRegion` is not an extension to `NSDate` | `NSDate` extensions in general represent the current date in various formats. Extensions also limit the class as you cannot add stored properties. The latter I need to store the regional data (calendar, time zone & locale). Cocoapods offers plenty `NSDate` extensions e.g. [SwiftDate](www.cocoapods.com/pods/SwiftDate).
Do not include an initialiser from string  | That is too complicated with all the different ways of notation in the world. It would have too little benefit next to the currently available convenience intialisers instead.
Do not attempt to mimic all properties and functions of the NSDateFormatter, NSDateComponents etc. E.g. NSDateFormatter's ``localizedStringFromDate`` or ``weekdaySymbols`` | Sometimes it is just easier to use the ``date`` property from ``DateInRegion`` instead.
Equation ``==`` is not for just the date value but for the whole object | Alike NSObject regulations; although objects do not have to be the same instance, they must contain equal properties

## Author

Jeroen Houtzager, pls contact me through GitHub

## Collaboration

if you would like to contribute:

- Fork
- Send pull requests
- Submit issues
- Challenge the design decisions
- Challenge coding
- Challenge anything!

These libs & authors inspired me to this code:

- [SwiftDate](https://github.com/malcommac/SwiftDate) from Daniele Margutti. In the mean time big parts of DateInRegion are incorporated in this library. Check it out!


## License

DateInRegion is available under the MIT license. See the LICENSE file for more info.


CHANGELOG.md
# Change Log

## [0.5.2](https://github.com/Hout/DateInRegion/tree/0.5.2) (2015-11-30)

- Bugfixes in testing and description
- Got Travis CI working

## [0.5.1](https://github.com/Hout/DateInRegion/tree/0.5.1) (2015-11-18)

- Renamed from JHDate to DateInRegion


## [0.5.0](https://github.com/Hout/DateInRegion/tree/0.5.0) (2015-11-08)

- Implemented ``Hashable & Equatable`` protocols
- Separate implementation files for various protocols
- Added weekend functions: ``nextWeekend(), previousWeekend, inWeekend`` incl all tests
- Made testing possible with other default regions, calendars
- Fixed leap month bug & added leap year function for Gregorian calendars
- In `EndOf` function, use 10 microseconds before the start of the next unit. This is the minimal difference where `NSDate.isEqualDate` returns a `false` value 


## [0.4.2](https://github.com/Hout/DateInRegion/tree/0.4.2) (2015-11-03)
**Implemented hotfix:**

- Readme doc badge adjusted

## [0.4.1](https://github.com/Hout/DateInRegion/tree/0.4.1) (2015-11-03)
**Implemented hotfix:**

- Podspec adjusted

## [0.4.0](https://github.com/Hout/DateInRegion/tree/0.4.0) (2015-11-03)
**Implemented enhancements:**

- Implement protocol NSSecureCoding [\#10](https://github.com/Hout/DateInRegion/issues/10)
- Implement protocol NSCopying [\#9](https://github.com/Hout/DateInRegion/issues/9)
  ...and got rid of that too. Made DateInRegion immutable.
- Add dateFormatter property [\#8](https://github.com/Hout/DateInRegion/issues/8)
- Document thread safety [\#6](https://github.com/Hout/DateInRegion/issues/6)
- Create README.playground [\#2](https://github.com/Hout/DateInRegion/issues/2)

**Closed issues:**

- IsEqual should compare entire object, not just the date [\#11](https://github.com/Hout/DateInRegion/issues/11)
- Add support for OSX [\#7](https://github.com/Hout/DateInRegion/issues/7)

## [0.3.1](https://github.com/Hout/DateInRegion/tree/0.3.1) (2015-11-01)
## [0.3.0](https://github.com/Hout/DateInRegion/tree/0.3.0) (2015-10-26)
**Implemented enhancements:**

- Split up in smaller and more comprehendable files [\#4](https://github.com/Hout/DateInRegion/issues/4)
- Complete access to all properties from NSDate and NSDateComponents [\#3](https://github.com/Hout/DateInRegion/issues/3)
- Include proper documentation [\#1](https://github.com/Hout/DateInRegion/issues/1)

## [0.2.1](https://github.com/Hout/DateInRegion/tree/0.2.1) (2015-10-25)
## [0.2.0](https://github.com/Hout/DateInRegion/tree/0.2.0) (2015-10-25)
## [v0.02](https://github.com/Hout/DateInRegion/tree/v0.02) (2015-10-19)
## [v0.01](https://github.com/Hout/DateInRegion/tree/v0.01) (2015-10-19)


\* *This Change Log was automatically generated by [github_changelog_generator](https://github.com/skywinder/Github-Changelog-Generator)*


Pod/Classes/DateRegion.swift
//
//  DateRegion.swift
//  Pods
//
//  Created by Jeroen Houtzager on 09/11/15.
//
//

public class DateRegion: Equatable {

    /// Calendar to interpret date values. You can alter the calendar to adjust the representation of date to your needs.
    ///
    public let calendar: NSCalendar!

    /// Time zone to interpret date values
    /// Because the time zone is part of calendar, this is a shortcut to that variable.
    /// You can alter the time zone to adjust the representation of date to your needs.
    ///
    public let timeZone: NSTimeZone!

    /// Locale to interpret date values
    /// Because the locale is part of calendar, this is a shortcut to that variable.
    /// You can alter the locale to adjust the representation of date to your needs.
    ///
    public let locale: NSLocale!
    
    /// Initialise with a calendar and/or a time zone
    ///
    /// - Parameters:
    ///     - calendar:   the calendar to work with to assign, default = the current calendar
    ///     - timeZone:   the time zone to work with, default is the default time zone
    ///     - locale:     the locale to work with, default is the current locale
    ///     - calendarID: the calendar ID to work with to assign, default = the current calendar
    ///     - timeZoneID: the time zone ID to work with, default is the default time zone
    ///     - localeID:   the locale ID to work with, default is the current locale
    ///     - region:     a region to copy
    ///
    /// - Note: parameters higher in the list take precedence over parameters lower in the list. E.g.
    ///     `DateRegion(locale: mylocale, localeID: "en_AU", region)` will copy region and set locale to mylocale, not `en_AU`.
    ///
    public init(
        calendarID: String = "",
        timeZoneID: String = "",
        localeID: String = "",
        calendar aCalendar: NSCalendar? = nil,
        timeZone aTimeZone: NSTimeZone? = nil,
        locale aLocale: NSLocale? = nil,
        region: DateRegion? = nil) {
            calendar = aCalendar ?? NSCalendar(calendarIdentifier: calendarID) ?? region?.calendar ?? NSCalendar.currentCalendar()
            timeZone = aTimeZone ?? NSTimeZone(abbreviation: timeZoneID) ?? NSTimeZone(name: timeZoneID) ?? region?.timeZone ?? NSTimeZone.defaultTimeZone()
            locale = aLocale ?? (localeID != "" ? NSLocale(localeIdentifier: localeID) : nil) ?? region?.locale ?? aCalendar?.locale ?? NSLocale.currentLocale()

            // Assign calendar fields
            calendar.timeZone = timeZone
            calendar.locale = locale
    }
    
    /// Today's date
    ///
    /// - Returns: the date of today at midnight (00:00) in the current calendar and default time zone.
    ///
    public func today() -> DateInRegion {
        let components = calendar.components([.Era, .Year, .Month, .Day, .Calendar, .TimeZone], fromDate: NSDate())
        let date = calendar.dateFromComponents(components)!
        return DateInRegion(region: self, date: date)
    }

    /// Yesterday's date
    ///
    /// - Returns: the date of yesterday at midnight (00:00) in the current calendar and default time zone.
    ///
    public func yesterday() -> DateInRegion {
        return (today() - 1.days)!
    }

    /// Tomorrow's date
    ///
    /// - Returns: the date of tomorrow at midnight (00:00) in the current calendar and default time zone.
    ///
    public func tomorrow() -> DateInRegion {
        return (today() + 1.days)!
    }

}

public func ==(left: DateRegion, right: DateRegion) -> Bool {
    if left.calendar.calendarIdentifier != right.calendar.calendarIdentifier {
        return false
    }

    if left.timeZone.secondsFromGMT != right.timeZone.secondsFromGMT {
        return false
    }

    if left.locale.localeIdentifier != right.locale.localeIdentifier  {
        return false
    }

    return true
}

extension DateRegion : Hashable {
    public var hashValue: Int {
        return calendar.hashValue ^ timeZone.hashValue ^ locale.hashValue
    }
}

extension DateRegion : CustomStringConvertible {
    public var description: String {
        let timeZoneAbbreviation = timeZone.abbreviation ?? ""
        return "\(calendar.calendarIdentifier); \(timeZone.name):\(timeZoneAbbreviation); \(locale.localeIdentifier)"
    }
}

Pod/Classes/DateInRegionDescription.swift
//
//  DateInRegionDescription.swift
//  Pods
//
//  Created by Jeroen Houtzager on 26/10/15.
//
//

import Foundation

// MARK: - CustomStringConvertable delegate

extension DateInRegion : CustomDebugStringConvertible {

    /// Returns a full description of the class
    ///
    public var description: String {
        return "\(self.toString()!); region: \(region)"
    }

    /// Returns a full debug description of the class
    ///
    public var debugDescription: String {
        var descriptor: [String] = []

        let formatter = NSDateFormatter()
        formatter.dateStyle = .LongStyle
        formatter.timeStyle = .LongStyle
        formatter.locale = self.locale
        formatter.calendar = self.calendar
        formatter.timeZone = self.timeZone
        descriptor.append(formatter.stringFromDate(self.date))

        formatter.timeZone = NSTimeZone(abbreviation: "UTC")
        descriptor.append("UTC\t\(formatter.stringFromDate(self.date))")

        descriptor.append("Calendar: \(calendar.calendarIdentifier)")
        descriptor.append("Time zone: \(timeZone.name)")
        descriptor.append("Locale: \(locale.localeIdentifier)")

        return descriptor.joinWithSeparator("\n")
    }
    
}




Pod/Classes/DateInRegionFormatterPort.swift
//
//  DateInRegionFormatterPort.swift
//  Pods
//
//  Created by Jeroen Houtzager on 01/11/15.
//
//

import Foundation

// MARK: - NSDateFormatter port

public extension DateInRegion {

    /// Returns a date representation of a given string interpreted using the receiver’s current settings.
    ///
    /// - Returns: A date representation of *string* interpreted using the receiver’s current settings.
    ///
    public func toString(format: String) -> String? {
        let formatter = NSDateFormatter()
        formatter.dateFormat = format
        formatter.locale = self.locale
        formatter.calendar = self.calendar
        formatter.timeZone = self.timeZone
        return formatter.stringFromDate(date)
    }

    public func toString(
        style: NSDateFormatterStyle? = nil,
        dateStyle: NSDateFormatterStyle? = nil,
        timeStyle: NSDateFormatterStyle? = nil) -> String? {

            let formatter = NSDateFormatter()
            formatter.dateStyle = style ?? dateStyle ?? .NoStyle
            formatter.timeStyle = style ?? timeStyle ?? .NoStyle
            if formatter.dateStyle == .NoStyle && formatter.timeStyle == .NoStyle {
                formatter.dateStyle = .MediumStyle
                formatter.timeStyle = .MediumStyle
            }
            formatter.locale = self.locale
            formatter.calendar = self.calendar
            formatter.timeZone = self.timeZone
            let string = formatter.stringFromDate(date)
            return string
    }

    /// Returns a relative date representation of a given string interpreted using the receiver’s current settings.
    /// If a date formatter uses relative date formatting, where possible it replaces the date component of its 
    /// output with a phrase—such as “today” or “tomorrow”—that indicates a relative date. The available phrases 
    /// depend on the locale for the date formatter; whereas, for dates in the future, English may only allow 
    /// “tomorrow,” French may allow “the day after the day after tomorrow,”.
    ///
    /// - Returns: A relative date representation of *string* interpreted using the receiver’s current settings.
    ///
    public func toRelativeString() -> String? {
        let formatter = NSDateFormatter()
        formatter.locale = self.locale
        formatter.calendar = self.calendar
        formatter.timeZone = self.timeZone
        formatter.dateStyle = .MediumStyle
        formatter.doesRelativeDateFormatting = true
        return formatter.stringFromDate(date)
    }
}

DateInRegion.playground/Contents.swift
/*:

# DateInRegion

DateInRegion is a wrapper around NSDate that exposes the properties of NSDateComponents, NSCalendar, NSTimeZone, NSLocale and NSDateFormatter. The best way to see it is as a localised (locale) moment (date) in a time zone (timeZone) within a calendrical system (calendar). The intention is to replace your occurrence of NSDate with DateInRegion and get the same functionality plus lots of local date/calendar/time zone/formatter goodies. Thus offering date functions with a flexibility that I was looking for when creating this library.

An extension is available for NSDate with which enables you to map an NSDate object to a DateInRegion object by adding the region.

### Features

- Use the object as an NSDate. I.e. as an absolute time.
- Offers many NSDate, NSCalendar, NSDateFormatter & NSDateComponent vars & methods
- Initialise a date with any combination of components
- DateInRegion is immutable, so thread safe. It contains a constructor to easily create new `DateInRegion` occurrences with some properties adjusted.
- Implements date addition and subtraction operators with date components. E.g. `date + 2.days`
- Default date is `NSDate()`
- Default calendar is `NSCalendar.currentCalendar()`
- Default time zone is `NSTimeZone.defaultTimeZone()`
- Default locale is `NSLocale.currentLocale()`
- Contains a date (NSDate), a calendar (NSCalendar), a locale (NSLocale) and a timeZone (NSTimeZone) property
- Implements the ``Equatable`` & ``Comparable`` protocols betwen dates with operators. E.g. `==, !=, <, >, <=, >=`
- Implements the ``Hashable`` protocol so the date can be used as a key in a Dictionary.

*/

import DateInRegion

// First create some regions
let local = DateRegion()
let netherlands = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "Europe?Amsterdam", localeID: "nl_NL")
let india = DateRegion(calendarID: NSCalendarIdentifierIndian, timeZoneID: "IST", localeID: "en_IN")
let china = DateRegion(calendarID: NSCalendarIdentifierChinese, timeZoneID: "Asia/Shanghai", localeID: "zn_CH")

// Initialisers
// Create a new date object with the current date & time alike NSDate()
let now = DateInRegion()

// Create a determined date
let newDate = DateInRegion(year: 2011, month: 2, day: 11)!

// Create a determined date in a different time zone
let newYork = DateRegion(timeZoneID: "EST")
DateInRegion(fromDate: newDate, hour: 14, region: newYork)!

// Mind that default values for DateInRegion(year etc) are taken from the reference date,
// which is 1 January 2001, 00:00:00.000 in your default time zone and against your current calendar.

// Week oriented initiailisations are also possible: first week of 2016:
let weekDate = DateInRegion(yearForWeekOfYear: 2016, weekOfYear: 1, weekday: 1)!
// In Europe this week starts in 2015 despite the year for the week that is 2016.
// That is because the Thursday of this week is in 2016 as specified by ISO 8601

// Or get date components
now.hour
now.day
now.month
now.year

// NSDate conversions
let nsDate = NSDate()
nsDate.inRegion(newYork).toString()
nsDate.inRegion(india).toString()
nsDate.inRegion(china).toString()


// Equations
// DateInRegion conforms to the Equatable protocol. I.e. you can compare with == for equality.
let newDate2 = DateInRegion(fromDate: newDate)
let newDate3 = DateInRegion(fromDate: newDate, hour: 9)!
newDate == newDate
newDate == newDate2
newDate == newDate3

// For equal date values, you should use x.isEqualToDate(y) note that this compares the moment not the region

// Comparisons
// DateInRegion conforms to the Comparable protocol. I.e. you can compare with <. <=, ==, >=, >
newDate > newDate2
newDate >= newDate2
newDate == newDate2
newDate != newDate2
newDate <= newDate2
newDate < newDate2

// Mind that comparisons takes the absolute time from the date property into account.
// regions (calendars, time zones, locales, have no effect on the comparison results.
let date1 = DateInRegion(year: 2000, month: 1, day: 1, hour: 14, region: netherlands)!
let date2 = (date1 + 1.hours)!
let date3 = (date2 - 2.weeks)!

// date1 = 14:00 UTC
// date2 = 15:00 UTC
date1 > date2
date1 < date2

let indiaDate1 = DateInRegion(fromDate: date1, region: india)

// indiaDate1 = 18:30 IST
// date1 = 14:00 UTC
// date2 = 9:00 CST
indiaDate1 > date2
indiaDate1 < date2

indiaDate1.isEqualToDate(date1)
// QED: same outcome!

// Collections
// Use as a key in a dictionary
var birthdays = [DateInRegion: String]()
birthdays[DateInRegion(year: 1997, month: 5, day: 7)!] = "Jerry"
birthdays[DateInRegion(year: 1999, month: 12, day: 14)!] = "Ken"
birthdays[DateInRegion(year: 1990, month: 12, day: 5)!] = "Sinterklaas"
birthdays.sort({ (a: (date: DateInRegion, String), b: (date: DateInRegion, String)) -> Bool in
    return a.date < b.date
})

// Display & string conversion
// DateInRegion conforms to the ConvertString protocol
indiaDate1.description

// DateInRegion conforms to the DebugConvertString protocol
indiaDate1.debugDescription

// Create some more regions to illustrate conversions & string handling
let tibetan = DateRegion(calendarID: NSCalendarIdentifierIndian, timeZoneID: "IST", localeID: "bo_IN")
let dubai = DateRegion(calendarID: NSCalendarIdentifierIslamic, timeZoneID: "GST", localeID: "ar_AE")
let newZealand = DateRegion(calendarID: NSCalendarIdentifierGregorian, localeID: "en_NZ", timeZoneID: "Pacific/Auckland")
let israel = DateRegion(calendarID: NSCalendarIdentifierHebrew, timeZoneID: "Asia/Jerusalem", localeID: "he_IL")
let magadan = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "Asia/Magadan", localeID: "ru_RU")
let thailand = DateRegion(calendarID: NSCalendarIdentifierBuddhist, timeZoneID: "Asia/Bangkok", localeID: "th_TH")
let japan = DateRegion(calendarID: NSCalendarIdentifierBuddhist, timeZoneID: "Asia/Tokyo", localeID: "ja_JP")
let unalaska = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "AKST", localeID: "en_US")
let utc = DateRegion(calendarID: NSCalendarIdentifierGregorian, timeZoneID: "UTC", localeID: "en_US_POSIX")


// Conversions
let unalaskaDate = now.inRegion(unalaska)
let netherlandsDate = now.inRegion(netherlands)
let newYorkDate = now.inRegion(newYork)
let indiaDate = now.inRegion(india)
let tibetanDate = now.inRegion(tibetan)
let dubaiDate = now.inRegion(dubai)
let israelDate = now.inRegion(israel)
let chinaDate = now.inRegion(china)
let newZealandDate = now.inRegion(newZealand)
let magadanDate = now.inRegion(magadan)
let japanDate = now.inRegion(japan)
let thailandDate = now.inRegion(thailand)
let utcDate = now.inRegion(utc)

// Or convert to local
let japanDate2 = DateInRegion(year: 2010, month: 4, day: 5, hour: 16, region: japan)!
let localDate = japanDate2.inLocalRegion()

// Now look in regional format
unalaskaDate.toString()
netherlandsDate.toString()
newYorkDate.toString()
indiaDate.toString()
tibetanDate.toString()
dubaiDate.toString()
israelDate.toString()
chinaDate.toString()
magadanDate.toString()
japanDate.toString()
thailandDate.toString()
newZealandDate.toString()
utcDate.toString()
japanDate2.toString()
localDate.toString()

// Other string formats: with styles
newYorkDate.toString(dateStyle: .LongStyle)
newYorkDate.toString(timeStyle: .LongStyle)
newYorkDate.toString(dateStyle: .ShortStyle, timeStyle: .MediumStyle)

// Other string formats: with Unicode abbreviations 
// http://www.unicode.org/reports/tr35/tr35-31/tr35-dates.html#Date_Format_Patterns
newYorkDate.toString("yy-MM-dd")
newYorkDate.toString("yyyy-DDD")
newYorkDate.toString("yy.ww.e")
newYorkDate.toString("HH:mm:ss.SSS")
newYorkDate.toString("hh:mm:ss a z")


